### ISSUE TITLE
Raspberry PI 3/4 support

### ISSUE THREAD

**Issue by circutrider21** [FlorenceOS/Sabaton#7]
I am currently working on a kernel of mine that uses sabaton and was wondering if there will be support raspberry pi 3/4.

**Comment by N00byEdge** [comment]
#6 is rpi3 support, rpi4 is planned to start once rpi3 support is finished.

**Comment by circutrider21** [comment]
Ok!

**Comment by circutrider21** [comment]
Although since the pull request seems a bit stale, I was wondering if I could continue from where @staticsaga stopped?

**Comment by N00byEdge** [comment]
I merged the pi3 code into master, works in a VM now (`zig build pi3`). I do have a pi3, and I would like to test it on there, but feel free to take on a pi4 port, or to improve the existing pi3 code.

**Comment by N00byEdge** [comment]
- [ ] Test on Pi3 HW
- [ ] Pi4 support

**Comment by circutrider21** [comment]
I too personally own a rpi3 but I also don't have a rpi4. I only asked out of curiosity. As for the testing, I'll start on it tonight.

**Comment by N00byEdge** [comment]
With this at the bottom of the config:

```
enable_uart=1
initramfs flork.elf 0x200000
```
Only getting a rainbow screen and a single uart null byte, so no dice yet

**Comment by circutrider21** [comment]
I got the same result with the rainbow screen, but I didn't see anything on the uart. Might be my diy adapter.

**Comment by N00byEdge** [comment]
Yeah, something is going wrong. Not working, even with all the EL2 code, relocation etc commented out, and halting on non-bsps, it's still not working. Feels like we're doing something wrong somewhere that only the HW cares about in regards to the uart init.

**Comment by circutrider21** [comment]
Have you tried commenting out the code that inits the uart.

**Comment by N00byEdge** [comment]
No, I have not. May be worth a try to see if that lets it continue on to the display.

**Comment by circutrider21** [comment]
Well, I tried that and it didn't do anything. Still getting the same old results...

**Comment by N00byEdge** [comment]
Yeah, I don't know what the next step to proceed would be, as I can't get any uart output and I don't have the ability to hook up jtag, so I'm kind of stuck at the moment. I'll take a look at using some existing code and comparing what works and not.

**Comment by circutrider21** [comment]
Sounds good!

**Comment by circutrider21** [comment]
Okay, the uart seems to work now that I have a new adapter, but the bootloader only prints 2 lines
`Allocating kernel memory`
`Bytes allocated for kernel: 000000000003A000`

**Comment by StaticSaga** [comment]
Sabaton in release modes doesn't print a lot of debugging information, switch build.zig to building with .Debug. You will get some more information about where it is failing.
I think that it is hanging somewhere in the mailbox code (responsible for setting the framebuffer and querying the memory available).
If you have Discord, I would join the FlorenceOS server. https://discord.gg/HpGsGjFG5S

**Comment by circutrider21** [comment]
I have discord, my username is SirNT (or devlooper) so I'll join the discord. As for building in debug mode, how would I do that (I am not familiar with zig)

**Comment by StaticSaga** [comment]
Replace the .ReleaseSmall in build.zig with .Debug

**Comment by ghost** [comment]
> I am currently working on a kernel of mine that uses sabaton and was wondering if there will be support raspberry pi 3/4.

Why so you can copy/paste the code? 😂👎
